src/roswire/ros2/node_manager.py

```diff
@@ -1,23 +1,26 @@
 # -*- coding: utf-8 -*-
 __all__ = ('ROS2NodeManager',)
 
-from typing import Iterator, Mapping
+from typing import Iterator
 import typing
 
 from loguru import logger
 import attr
 
 from .node import ROS2Node
+from .state import ROS2StateProbe
 from .. import exceptions as exc
+from ..interface import NodeManager
 
 if typing.TYPE_CHECKING:
     from ..app import AppInstance
 
 
 @attr.s(frozen=True, auto_attribs=True, slots=True)
-class ROS2NodeManager(Mapping[str, ROS2Node]):
+class ROS2NodeManager(NodeManager):
     """Provides an interface for interacting with ROS2 nodes."""
-    app_instance: 'AppInstance'
+    app_instance: 'AppInstance' = attr.ib()
+    state: 'ROS2StateProbe' = ROS2StateProbe.for_app_instance(app_instance)
```

Comment:
`state` doesn't achieve what you're after here (and will likely lead to a crash at runtime). In its current form, this code tries to create a class variable `state` using an uninstantiated instance variable `app_instance`.

Implementing this requires a more advanced use of attrs. To implement the `NodeManager` interface, you do indeed want to have a `state` attribute, but you want to instantiate the state probe in the constructor. However, attrs doesn't have a conventional `__init__` constructor! Fortunately, attrs calls a `__attrs_post_init__` method after it has initialised an object but before that object is given to any potential caller.

The trick here is to declare the `state` attribute using `attr.ib(init=False)`, and to defer its initialisation to `__attrs_post_init__`. Normally, you would just be able to do `self.state = ROS2StateProbe.for_app_instance(app_instance)` inside `__attrs_post_init__`. However, here we have made `ROS2NodeManager` a frozen (i.e., immutable) class. That means we can't directly modify `self`! Thankfully, we can use a trick to set the value of the `state` attribute inside `__attrs_post_init__`: https://github.com/ChrisTimperley/roswire/blob/9e7cffa35693d40bdf31a8ddb757b8ef1b8f6c57/src/roswire/ros2/ros2.py#L27-L35

Reply:
Having said all of this, I realise that `state` is not a `ROS2StateProbe` in the `NodeManager` interface. `state` should be implemented as a `@property` that uses the state probe behind the scenes to obtain the state of the system.

In that case, the probe should become `_state_probe`, but all of the other comments above still apply.